ssl/s3_lib.c

```diff
@@ -3371,6 +3371,10 @@ void ssl3_free(SSL *s)
     OPENSSL_free(sc->s3.alpn_selected);
     OPENSSL_free(sc->s3.alpn_proposed);
 
+#ifndef OPENSSL_NO_PSK
+    OPENSSL_free(sc->s3.tmp.psk);
```

Comment:
I am wondering why we have never noticed this before. Surely this leak would occur in normal operation???

Reply:
I had noticed it with memory error injection (in the client) but it looks likely that the
ssl_generate_master_secret can be skipped due to a parse error...

Reply:
```
Indirect leak of 20 byte(s) in 1 object(s) allocated from:
    #0 0x7f19ca8b89cf in __interceptor_malloc ../../../../src/libsanitizer/asan/asan_malloc_linux.cpp:69
    #1 0x7f19ca356605 in CRYPTO_malloc crypto/mem.c:221
    #2 0x7f19ca356605 in CRYPTO_malloc crypto/mem.c:198
    #3 0x7f19ca3823d7 in CRYPTO_memdup crypto/o_str.c:64
    #4 0x7f19ca7573d2 in tls_construct_cke_psk_preamble ssl/statem/statem_clnt.c:2932
    #5 0x7f19ca7582fe in tls_construct_client_key_exchange ssl/statem/statem_clnt.c:3334
    #6 0x7f19ca753309 in write_state_machine ssl/statem/statem.c:843
    #7 0x7f19ca753309 in state_machine ssl/statem/statem.c:443
    #8 0x7f19ca71cf8f in SSL_do_handshake ssl/ssl_lib.c:3757
    #9 0x55c2e81d30c7 in create_bare_ssl_connection test/ssltestlib.c:881
    #10 0x55c2e81d36b6 in create_ssl_connection test/ssltestlib.c:956
    #11 0x55c2e81cf687 in mtu_test test/dtls_mtu_test.c:71
    #12 0x55c2e81cfdfe in run_mtu_tests test/dtls_mtu_test.c:174
    #13 0x55c2e81d44ee in run_tests test/testutil/driver.c:205
    #14 0x55c2e81cec5a in main test/testutil/main.c:51
    #15 0x7f19c9e461c9 in __libc_start_call_main ../sysdeps/nptl/libc_start_call_main.h:58
```